consensus/parlia/snapshot.go

```diff
@@ -346,17 +394,20 @@ func (s *Snapshot) validators() []common.Address {
 	return validators
 }
 
+// lastBlockInOneTurn returns if the block at height `blockNumber` is the last block in current turn.
+func (s *Snapshot) lastBlockInOneTurn(blockNumber uint64) bool {
+	return (blockNumber+1)%uint64(s.TurnLength) == 0
+}
+
 // inturn returns if a validator at a given block height is in-turn or not.
 func (s *Snapshot) inturn(validator common.Address) bool {
-	validators := s.validators()
-	offset := (s.Number + 1) % uint64(len(validators))
-	return validators[offset] == validator
+	return s.inturnValidator() == validator
 }
 
 // inturnValidator returns the validator at a given block height.
 func (s *Snapshot) inturnValidator() common.Address {
 	validators := s.validators()
-	offset := (s.Number + 1) % uint64(len(validators))
+	offset := (s.Number + 1) / uint64(s.TurnLength) % uint64(len(validators))
```

Comment:
`(s.Number + s.TurnLength) / uint64(s.TurnLength)`?

Reply:
> `(s.Number + s.TurnLength) / uint64(s.TurnLength)`?

when at height `s.Number + 1` ,  we can get snapshot with s.Number.
the function name is not quite clear.

Reply:
> > `(s.Number + s.TurnLength) / uint64(s.TurnLength)`?
> 
> when at height `s.Number + 1` , we can get snapshot with s.Number. the function name is not quite clear. I will update the comments of the function

Reply:
(s.Number + s.TurnLength) == (s.Number + 1 + (s.TurnLength -1))

Reply:
> (s.Number + s.TurnLength) == (s.Number + 1 + (s.TurnLength -1))

not enable the first validator after switch go generate s.TurnLength blocks